### PR TITLE
Weighted poisson integration

### PR DESCRIPTION
**Pull request by jhenin**
Implements weighted Poisson integration of free energy gradients.
Also fixes a small error in Neumann boundary conditions for the existing unweighted case.

TODO:

- [x] revert formatting changes (braces and indents)
- [x] user flag to enable/disable
- [x] Documentation
- [x] Fix failing tests due to boundary condition fixes in unweighted Poisson
- [x] Add test for new feature, weighted Poisson

### CODE REVIEW

File: namd/tests/library/018_pathCV/AutoDiff/test.pmf [hunk 1]
```diff
@@ -2,1220 +2,1220 @@
...
- -2.00000000000000e-01  1.10000000000000e+00   7.84861755647861e-02
- -2.00000000000000e-01  1.15000000000000e+00   7.81055663777811e-02
- -2.00000000000000e-01  1.20000000000000e+00   7.77858043866154e-02
- -2.00000000000000e-01  1.25000000000000e+00   7.75232734973943e-02
- -2.00000000000000e-01  1.30000000000000e+00   7.73142286913480e-02
- -2.00000000000000e-01  1.35000000000000e+00   7.71553039111600e-02
- -2.00000000000000e-01  1.40000000000000e+00   7.70437678236428e-02
- -2.00000000000000e-01  1.45000000000000e+00   7.69776442679124e-02
- -2.00000000000000e-01  1.50000000000000e+00   7.69557395004421e-02
+ -2.00000000000000e-01 -5.00000000000000e-01   5.66747557569237e-02
```
Comment (HanatoK) [inline namd/tests/library/018_pathCV/AutoDiff/test.pmf]: Why are the test data changed here? I know this PR fixes the boundary condition in the integrator but `018_pathCV` does not use any periodic CVs.

**Comment by HanatoK** [comment]
When restarting `000_eulerangles_ext_abf2d_euler_weighted`, NAMD crashes. The GDB backtrace:
```
#0  0x00007ffff763dd3c in __pthread_kill_implementation () from /lib64/libc.so.6
#1  0x00007ffff75e27b6 in raise () from /lib64/libc.so.6
#2  0x00007ffff75c934b in abort () from /lib64/libc.so.6
#3  0x00007ffff7889972 in std::__glibcxx_assert_fail(char const*, int, char const*, char const*) () from /lib64/libstdc++.so.6
#4  0x00000000009b1cd7 in std::vector<double, std::allocator<double> >::operator[] (this=0x1d85b08, __n=0) at /usr/include/c++/15/bits/stl_vector.h:1263
#5  0x00000000013ba889 in colvargrid_integrate::update_weighted_div_local (this=0x1d85770, ix0=std::vector of length 2, capacity 2 = {...}) at colvars/src/colvargrid_integrate.cpp:344
#6  0x00000000013b8029 in colvargrid_integrate::set_weighted_div (this=0x1d85770) at colvars/src/colvargrid_integrate.cpp:151
#7  0x00000000012b8869 in colvargrid_integrate::set_div (this=0x1d85770) at colvars/src/colvargrid_integrate.h:50
#8  0x00000000012c48fb in colvarbias_abf::read_state_data_template_<std::basic_istream<char, std::char_traits<char> > > (this=0x1d5a880, is=...) at colvars/src/colvarbias_abf.cpp:968
#9  0x00000000012b66ab in colvarbias_abf::read_state_data (this=0x1d5a880, is=...) at colvars/src/colvarbias_abf.cpp:1017
#10 0x00000000012a9a2a in colvarbias::read_state_template_<std::basic_istream<char, std::char_traits<char> > > (this=0x1d5a880, is=...) at colvars/src/colvarbias.cpp:641
#11 0x00000000012a43fb in colvarbias::read_state (this=0x1d5a880, is=...) at colvars/src/colvarbias.cpp:665
#12 0x00000000013d2c5e in colvarmodule::read_objects_state (this=0x1b92970, is=...) at colvars/src/colvarmodule.cpp:1811
#13 0x00000000013e7d2f in colvarmodule::read_state_template_<std::basic_istream<char, std::char_traits<char> > > (this=0x1b92970, is=...) at colvars/src/colvarmodule.cpp:1724
#14 0x00000000013d25b9 in colvarmodule::read_state (this=0x1b92970, is=...) at colvars/src/colvarmodule.cpp:1732
#15 0x00000000013d11fe in colvarmodule::setup_input (this=0x1b92970) at colvars/src/colvarmodule.cpp:1562
#16 0x00000000014274e0 in cvscript_cv_load (pobj=0x1b4b140, objc=3, objv=0x7fffffffb9b0) at colvars/src/colvarscript_commands.h:484
#17 0x000000000141c9c1 in colvarscript::run (this=0x1b4b140, objc=3, objv=0x7fffffffb9b0) at colvars/src/colvarscript.cpp:439
#18 0x000000000141e87f in tcl_run_colvarscript_command (my_interp=0x1a75410, objc=3, objv=0x1a79790) at colvars/src/colvarscript.cpp:740
#19 0x00007ffff7de3ab8 in TclNRRunCallbacks () from /lib64/libtcl8.6.so
#20 0x00007ffff7dea234 in ?? () from /lib64/libtcl8.6.so
#21 0x00007ffff7e90587 in Tcl_FSEvalFileEx () from /lib64/libtcl8.6.so
#22 0x00007ffff7e90807 in Tcl_EvalFile () from /lib64/libtcl8.6.so
#23 0x0000000001121e77 in ScriptTcl::load (this=0x1a742b0, scriptFile=0x7fffffffd459 "test.restart.namd") at src/ScriptTcl.C:2423
#24 0x00000000009c463c in after_backend_init (argc=2, argv=0x7fffffffce68) at src/mainfunc.C:181
#25 0x00000000009c3f88 in main (argc=3, argv=0x7fffffffce68) at src/mainfunc.C:52
```

File: namd/tests/library/018_pathCV/AutoDiff/test.pmf (hunk 1 above)
Comment (jhenin) [reply namd/tests/library/018_pathCV/AutoDiff/test.pmf]: The fix applies to Neumann (non-periodic) boundary conditions.

File: namd/tests/library/000_distance-grid_abf-customgrid/test.in [hunk 2]
```diff
@@ -31,6 +31,6 @@ abf {
     grid {
         width 0.2
         lowerBoundary 1.0
-        upperBoundary 5.0
+        upperBoundary 6.0
```
[the remark this answers is not in the document]
Comment (jhenin) [reply namd/tests/library/000_distance-grid_abf-customgrid/test.in]: This ensures that the custom grid has a different number of points than the default one based on the colvar's parameters. Because there was an accidental match in the previous test, some errors went unnoticed.

File: misc_interfaces/stubs/colvarproxy_stub.cpp [hunk 3]
```diff
@@ -74,7 +74,7 @@ bool colvarproxy_stub::total_forces_enabled() const
 
 bool colvarproxy_stub::total_forces_same_step() const
 {
-  return total_force_requested;
+  return true;
```
[the remark this answers is not in the document]
Comment (jhenin) [reply misc_interfaces/stubs/colvarproxy_stub.cpp]: I think so, but @giacomofiorin may have a different opinion. I didn't see a rationale for that flag changing based on total_force_requested, i.e. the fact that it's true does not indicate that total forces are enabled or requested or even available.

File: src/colvargrid_integrate.cpp [hunk 4]
```diff
@@ -382,313 +521,851 @@ void colvargrid_integrate::atimes(const std::vector<cvm::real> &A, std::vector<c
...
+  if (weighted) {
+    std::vector<int> max_position;
+    std::vector<int> min_position;
+    sorted_counts = {};
+
+    for (std::vector<int> ix = gradients->new_index(); gradients->index_ok(ix);
+         gradients->incr(ix)) {
+      size_t count = gradients->samples->value(ix);
+      if (count > 0) {
+        insert_into_sorted_list<size_t>(sorted_counts, count);
```
[the remark this answers is not in the document]
Comment (jhenin) [reply src/colvargrid_integrate.cpp]: Agreed, thanks!

File: src/colvargrid_integrate.cpp [hunk 5]
```diff
@@ -146,92 +189,187 @@ void colvargrid_integrate::update_div_neighbors(const std::vector<int> &ix0)
...
+    for (size_t i = 0; i < nd; i++) {
+      surrounding_point_coordinates[i] += surrounding_point_relative_position[i];
+    }
+
+    gradients->wrap_detect_edge(surrounding_point_coordinates);
+    get_regularized_grad(gradient_at_surrounding_point, surrounding_point_coordinates);
+    cvm::real weight = regularized_weights[gradients->address(surrounding_point_coordinates)];
+
+    for (size_t i = 0; i < nd; i++) {
+      div_at_point += pow(-1, surrounding_point_relative_position[i] + 1) *
```
Comment (HanatoK) [inline src/colvargrid_integrate.cpp]: (-1)^n could be either -1 or 1 depending on whether n is an even number or an odd number, so this can be simply calculated without calling `pow`.

File: src/colvargrid_integrate.cpp [hunk 6]
```diff
@@ -146,92 +189,187 @@ void colvargrid_integrate::update_div_neighbors(const std::vector<int> &ix0)
...
   }
 }
 
 
+void colvargrid_integrate::prepare_divergence_stencils()
+{
+  surrounding_points_relative_positions.clear();
+  size_t n_combinations = pow(2, nd);
+  for (size_t i = 0; i < n_combinations; i++) {
+    surrounding_points_relative_positions.push_back(convert_base_two(i, nd));
```
Comment (HanatoK) [inline src/colvargrid_integrate.cpp]: I feel that the calls to `convert_base_two` is unnecessary because an integer has already all its bits itself. `convert_base_two` simply checks the bits of `i` and stores them into a vector of `nd` size. But maybe I am wrong, and I am not sure how to optimize this.

File: src/colvargrid_integrate.h [hunk 7]
```diff
@@ -1,78 +1,195 @@
...
+  void prepare_calculations();
 
   /// Array holding divergence + boundary terms (modified Neumann) if not periodic
+  std::vector<cvm::real> laplacian_matrix_test;
+
+protected:
+  bool weighted = false;
+  bool precompute = true;
+
+  colvar_grid_scalar* computation_grid;
```
[the remark this answers is not in the document]
Comment (jhenin) [reply src/colvargrid_integrate.h]: Thanks, since this is not managing any memory, we'll keep the raw pointer for simplicity but initialize to nullptr.

File: src/colvargrid_integrate.cpp [hunk 8]
```diff
@@ -146,92 +189,187 @@ void colvargrid_integrate::update_div_neighbors(const std::vector<int> &ix0)
...
+  }
+}
+
+/// Updates the divergence at the point ix0
+void colvargrid_integrate::update_weighted_div_local(const std::vector<int> &ix0)
+{
+  const size_t linear_index = computation_grid->address(ix0);
+  cvm::real div_at_point = 0;
+  for (std::vector<int> surrounding_point_relative_position :
```
[the remark this answers is not in the document]
Comment (jhenin) [reply src/colvargrid_integrate.cpp]: Done.
